### DISCUSSION TITLE
How to insert scss inside shadowRoot Web Components

### DISCUSSION THREAD

**Discussion by VelynnXV** [parcel-bundler/parcel#5847] [Q&A] [answered]
Is there an easy way to insert my scss file into my `template.innerHTML`? 
The purpose being encapsulation of this specific style to be effective only for this component.

I am using `"parcel-bundler": "^1.12.4"`

**How can I import the scss without it affecting the whole app + insert it into my web component template only?** 

I tried to place `<link rel="stylesheet" href="../scss/animeCardStyle.scss">` inside the innerHTML of my template, without any results.

This is my component
```js
// how do i include sass in template
import animeCardStyle from '../scss/animeCardStyle.scss';

const template = document.createElement('template');
template.innerHTML = `
    <div class="anime-card">
    <div>
        <img />
        <h3></h3>
        <div class="info">
            <p>
                <slot name="anime-description" >
            </p>
        </div>
    </div>
    </div>
`;

class AnimeCard extends HTMLElement{
    constructor(){
        super() // calling the constuctor of HTMLElement

        // ShadowDom
            //An important aspect of web components is encapsulation
            //being able to keep the markup structure, style, and behavior hidden and separate
            //from other code on the page so that different parts do not clash, and the code can 
            //be kept nice and clean. The Shadow DOM API is a key part of this, providing a way to 
            //attach a hidden separated DOM to an element

        this.attachShadow({mode:'open'})
        this.shadowRoot.appendChild(template.content.cloneNode(true))
        this.shadowRoot.querySelector('h3').innerHTML = this.getAttribute('anime-title')
        this.shadowRoot.querySelector('img').src = this.getAttribute('avatar')
        this.shadowRoot.querySelector('img').width = "300"

    }

    // Custom element lifecycle methods

    // constructor() : Called when an instance of the element is created or upgraded
    // connectedCallback() : Called everytime an element is inserted in DOM -> useful for setting event listeners
    // disconnectedCallback() : Called everytime an element is removed from DOM
    // attributeChangedCallback(attrName,oldValue,newValue) : Called when an attribute is added, removed, upgraded or replaced.
    something() {
        
      }

    // connectedCallback() {
    //     this.shadowRoot.querySelector('#elem').addEventListener('click', () => this.something());
    //   }
    
    //   disconnectedCallback() {
    //     this.shadowRoot.querySelector('#elem').removeEventListener();
    //   }
}
window.customElements.define('anime-card',AnimeCard)


const animeCard = async () => {

    let response = await fetch('https://kitsu.io/api/edge/anime').then(function (res) {
        return res.json()
    });

    let animes = response
    console.log(animes.data)


    let template = "";

    for (let index = 0; index < animes.data.length; index++) {
        let anime = animes.data[index];

        let anime_img = '';
        if (anime.attributes.coverImage != null) {
            anime_img = anime.attributes.coverImage.original;
        }
        let anime_title = '';
        if (anime.attributes.coverImage != null) {
            anime_title = anime.attributes.canonicalTitle;
        }
        let anime_description = '';
        if (anime.attributes.coverImage != null) {
            anime_description = anime.attributes.description;
        }

        template += `
                <anime-card avatar="${anime_img}" anime-title="${anime_title}">
                    <div slot="anime-description">${anime_description}</div>
                </anime-card>           
            `

    }

    return template
}

export default animeCard
```

My package.json

```json
{
  "name": "mypackage",
  "version": "1.0.0",
  "description": "This is my package",
  "main": "index.js",
  "scripts": {
    "dev": "npm run clean && parcel public/index.html --out-dir development -p 3000",
    "build": "rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./",
    "clean": "rimraf ./development && rimraf ./.cache"
  },
  "author": "me",
  "license": "UNLICENSED",
  "devDependencies": {
    "@babel/core": "^7.12.16",
    "@babel/plugin-transform-runtime": "^7.12.15",
    "@babel/runtime-corejs2": "^7.12.13",
    "@babel/runtime-corejs3": "^7.12.13",
    "parcel-bundler": "^1.12.4",
    "sass": "^1.32.7",
    "sass-to-string": "^1.4.2"
  }
}

```

**Comment by mischnic** [comment]
With Parcel 2, you get import the contents of the SCSS file with `bundle-text:`: https://v2.parceljs.org/configuration/plugin-configuration/#predefined-(offical)-named-pipelines

**Reply by VelynnXV** [reply]
Hey, It's the first time I use parcel sorry for my inexperience.
 
I replaced `import animeCardStyle from '../scss/animeCardStyle.scss';` by  `import animeCardStyle from "bundle-text:../scss/animeCardStyle.scss";` and i get the error on build `Cannot resolve dependency 'bundle-text:../scss/animeCardStyle.scss'`.

 Am i missing anything in my package.json? Thank you so much.

**Reply by mischnic** [reply]
> Parcel 2

You're probably using Parcel 1. Do `yarn add parcel@nightly` or `yarn add parcel@next` instead

**Reply by VelynnXV** [reply]
I did `npm install -D parcel@next` to install it,  here is my package.json

```json
{
  "name": "mypackage",
  "version": "1.0.0",
  "description": "This is my package, using Parcel. Parcel is a web application bundler, differentiated by its developer experience. It offers blazing fast performance utilizing multicore processing, and requires zero configuration.",
  "main": "index.js",
  "scripts": {
    "dev": "npm run clean && parcel public/index.html --out-dir development -p 3000",
    "build": "rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./",
    "clean": "rimraf ./development && rimraf ./.cache"
  },
  "author": "me",
  "license": "UNLICENSED",
  "devDependencies": {
    "@babel/core": "^7.12.16",
    "@babel/plugin-transform-runtime": "^7.12.15",
    "@babel/runtime-corejs2": "^7.12.13",
    "parcel": "^2.0.0-beta.1",
    "parcel-bundler": "^1.12.4",
    "postcss": "^8.2.6",
    "postcss-modules": "^4.0.0",
    "postcss-uncss": "^0.17.0",
    "sass": "^1.32.7",
    "uncss": "^0.17.3"
  },
  "dependencies": {}
}

```

When i run `npm run build ` i have the error : 

> > mypackage@1.0.0 build /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow
> > rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./
> 
> /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/core/lib/serializer.js:22
>     throw new Error('Name already registered with serializer');
>     ^
> 
> Error: Name already registered with serializer
>     at registerSerializableClass (/Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/core/lib/serializer.js:22:11)
>     at Object.<anonymous> (/Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/package-manager/node_modules/@parcel/fs/lib/NodeFS.js:126:37)
>     at Module._compile (internal/modules/cjs/loader.js:1063:30)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
>     at Module.load (internal/modules/cjs/loader.js:928:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:769:14)
>     at Module.require (internal/modules/cjs/loader.js:952:19)
>     at require (internal/modules/cjs/helpers.js:88:18)
>     at Object.<anonymous> (/Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/package-manager/node_modules/@parcel/fs/lib/index.js:28:15)
>     at Module._compile (internal/modules/cjs/loader.js:1063:30)
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! mypackage@1.0.0 build: `rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the mypackage@1.0.0 build script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/masyaf/.npm/_logs/2021-02-16T20_49_28_610Z-debug.log

**Reply by mischnic** [reply]
You should definitely remove the `parcel-bundler` dependency. https://github.com/parcel-bundler/parcel/issues/4064

**Reply by VelynnXV** [reply]
@mischnic i deleted package-lock.json, removed `parcel-bundler` dependency and did npm install again, the installation went without failures but when i run npm run build, I get 

> masyaf@MacBook-Pro-de-Velynn Front End Workflow % npm run build
> 
> > mypackage@1.0.0 build /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow
> > rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./
> 
> error: unknown option '--out-dir'
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! mypackage@1.0.0 build: `rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the mypackage@1.0.0 build script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/masyaf/.npm/_logs/2021-02-16T21_08_55_646Z-debug.log


Here is the full output: 

> 
> masyaf@MacBook-Pro-de-Velynn Front End Workflow % npm install  
> 
> > fsevents@1.2.13 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/fsevents
> > node install.js
> 
>   SOLINK_MODULE(target) Release/.node
>   CXX(target) Release/obj.target/fse/fsevents.o
>   SOLINK_MODULE(target) Release/fse.node
> 
> > @parcel/watcher@2.0.0-alpha.10 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/core/node_modules/@parcel/watcher
> > node-gyp-build
> 
> 
> > @parcel/watcher@2.0.0-alpha.10 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/package-manager/node_modules/@parcel/watcher
> > node-gyp-build
> 
> 
> > @parcel/source-map@2.0.0-alpha.4.13 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/source-map
> > node-gyp-build
> 
> 
> > @parcel/watcher@2.0.0-alpha.10 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/transformer-sass/node_modules/@parcel/watcher
> > node-gyp-build
> 
> 
> > @parcel/watcher@2.0.0-alpha.10 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/parcel/node_modules/@parcel/watcher
> > node-gyp-build
> 
> 
> > core-js@3.8.3 postinstall /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/transformer-babel/node_modules/core-js
> > node -e "try{require('./postinstall')}catch(e){}"
> 
> Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!
> 
> The project needs your help! Please consider supporting of core-js on Open Collective or Patreon: 
> > https://opencollective.com/core-js 
> > https://www.patreon.com/zloirock 
> 
> Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)
> 
> 
> > core-js@2.6.12 postinstall /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/core-js
> > node -e "try{require('./postinstall')}catch(e){}"
> 
> 
> > ejs@2.7.4 postinstall /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/ejs
> > node ./postinstall.js
> 
> Thank you for installing EJS: built with the Jake JavaScript build tool (https://jakejs.com/)
> 
> npm WARN mypackage@1.0.0 No repository field.
> 
> added 1315 packages from 740 contributors and audited 1315 packages in 32.068s
> 
> 61 packages are looking for funding
> masyaf@MacBook-Pro-de-Velynn Front End Workflow % npm install
> npm WARN deprecated core-js@2.6.12: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
> npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
> npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
> 
> > @parcel/source-map@2.0.0-alpha.4.21 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/source-map
> > node-gyp-build
> 
> 
> > @parcel/watcher@2.0.0-alpha.10 install /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/watcher
> > node-gyp-build
> 
> 
> > core-js@3.8.3 postinstall /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/@parcel/transformer-babel/node_modules/core-js
> > node -e "try{require('./postinstall')}catch(e){}"
> 
> 
> > core-js@2.6.12 postinstall /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/core-js
> > node -e "try{require('./postinstall')}catch(e){}"
> 
> 
> > ejs@2.7.4 postinstall /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow/node_modules/ejs
> > node ./postinstall.js
> 
> Thank you for installing EJS: built with the Jake JavaScript build tool (https://jakejs.com/)
> 
> npm notice created a lockfile as package-lock.json. You should commit this file.
> npm WARN mypackage@1.0.0 No repository field.
> 
> added 607 packages from 704 contributors and audited 607 packages in 17.881s
> 
> 90 packages are looking for funding
>   run `npm fund` for details
> 
> found 0 vulnerabilities
> 
> masyaf@MacBook-Pro-de-Velynn Front End Workflow % npm run build
> 
> > mypackage@1.0.0 build /Users/masyaf/Documents/Volt/Lynn Corp./Inustry/Front End Workflow
> > rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./
> 
> error: unknown option '--out-dir'
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! mypackage@1.0.0 build: `rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the mypackage@1.0.0 build script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/masyaf/.npm/_logs/2021-02-16T21_08_26_189Z-debug.log

**Comment by VelynnXV** [comment, marked as answer]
@mischnic Hey! it works! I had to change `--out-dir` to `--dist-dir` to run my app.

Here is the package.json :

```json
{
  "name": "mypackage",
  "version": "1.0.0",
  "description": "This is my package, using Parcel. Parcel is a web application bundler, differentiated by its developer experience. It offers blazing fast performance utilizing multicore processing, and requires zero configuration.",
  "main": "index.js",
  "scripts": {
    "dev": "npm run clean && parcel public/index.html --dist-dir development -p 3000",
    "build": "rimraf ./dist && parcel build public/*.html --out-dir dist --public-url ./",
    "clean": "rimraf ./development && rimraf ./.cache"
  },
  "author": "me",
  "license": "UNLICENSED",
  "devDependencies": {
    "@babel/core": "^7.12.16",
    "@babel/plugin-transform-runtime": "^7.12.15",
    "@babel/runtime-corejs2": "^7.12.13",
    "@parcel/optimizer-htmlnano": "^2.0.0-nightly.598",
    "@parcel/packager-css": "^2.0.0-nightly.598",
    "@parcel/packager-html": "^2.0.0-nightly.598",
    "@parcel/transformer-css": "^2.0.0-nightly.598",
    "@parcel/transformer-html": "^2.0.0-nightly.598",
    "@parcel/transformer-inline-string": "^2.0.0-nightly.598",
    "@parcel/transformer-postcss": "^2.0.0-nightly.598",
    "@parcel/transformer-posthtml": "^2.0.0-nightly.598",
    "@parcel/transformer-sass": "^2.0.0-nightly.598",
    "parcel": "^2.0.0-beta.1",
    "postcss": "^8.2.6",
    "sass": "^1.32.7"
  },
  "dependencies": {}
}

```

The code to make the encapsulation work 

```js
        let shadow = this.attachShadow({ mode: "open" });
        let style = document.createElement("style");
        style.textContent = mystyle;
        shadow.appendChild(style);
```

Dont forget to delete `package-lock.json` when you pass from parcel 1 to 2 using `npm install parcel@next`
